src/silentpayments.rs

```diff
@@ -0,0 +1,546 @@
+
+        Self(ffi::SilentpaymentsRecipient::new(
+            &copy_to_ffi_pubkey(scan_pubkey),
+            &copy_to_ffi_pubkey(spend_pubkey),
+            index
+        ))
+    }
+}
+
+/// Sender Output creation errors
+#[derive(Debug, Clone, Copy, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub enum SenderOutputCreationError {
+    /// Unexpected failures
+    Failure,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for SenderOutputCreationError {}
+
+impl fmt::Display for SenderOutputCreationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        match self {
+            SenderOutputCreationError::Failure => write!(f, "Failed to create silent payments outputs"),
+        }
+    }
+}
+
+/// Create Silent Payment outputs for recipient(s).
+pub fn silentpayments_sender_create_outputs<C: Verification>(
+    secp: &Secp256k1<C>,
+    recipients: &mut [&SilentpaymentsRecipient],
+    smallest_outpoint: &[u8; 36],
+    taproot_seckeys: Option<&[&Keypair]>,
+    plain_seckeys: Option<&[&SecretKey]>,
+) -> Result<Vec<XOnlyPublicKey>, SenderOutputCreationError> {
+    let cx = secp.ctx().as_ptr();
+    let n_tx_outputs = recipients.len();
+
+    unsafe {
+        let mut out_pubkeys = vec![ffi::XOnlyPublicKey::new(); n_tx_outputs];
+        let mut out_pubkeys_ptrs: Vec<_> = out_pubkeys.iter_mut().map(|k| k as *mut _).collect();
+
+        for (i, pubkey) in out_pubkeys.iter_mut().enumerate() {
+            out_pubkeys_ptrs[i] = pubkey as *mut _;
+        }
+
+        let ffi_recipients_ptrs: &mut [*const ffi::SilentpaymentsRecipient] =
+                    transmute::<&mut [&SilentpaymentsRecipient], &mut [*const ffi::SilentpaymentsRecipient]>(recipients);
+
+        let (ffi_taproot_seckeys, n_taproot_seckeys) = match taproot_seckeys {
+            Some(keys) => {
+                let ffi_keys: &[*const ffi::Keypair] = transmute::<&[&Keypair], &[*const ffi::Keypair]>(taproot_seckeys.unwrap());
+                (ffi_keys.as_c_ptr(), keys.len())
+            }
+            None => (core::ptr::null(), 0),
+        };
+
+        let (ffi_plain_seckeys, n_plain_seckeys) = match plain_seckeys {
+            Some(keys) => {
+                let ffi_keys: &[*const u8] = transmute::<&[&SecretKey], &[*const u8]>(plain_seckeys.unwrap());
+                (ffi_keys.as_c_ptr(), keys.len())
+            }
+            None => (core::ptr::null(), 0),
+        };
+
+        let res = secp256k1_silentpayments_sender_create_outputs(
+            cx,
+            out_pubkeys_ptrs.as_mut_ptr(),
+            ffi_recipients_ptrs.as_mut_ptr(),
+            n_tx_outputs,
+            smallest_outpoint.as_ptr(),
+            ffi_taproot_seckeys,
+            n_taproot_seckeys,
+            ffi_plain_seckeys,
+            n_plain_seckeys,
+        );
+
+        if res == 1 {
+            Ok(out_pubkeys.into_iter().map(XOnlyPublicKey::from).collect())
```

Comment:
This can be optimized if `XOnlyPublicKey` is `#[repr(transparent)]`. (using [Vec::from_raw_parts](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts))

Reply:
Thanks. Done in ac90bf3093977945f19efbf2c6e088efb7b193a8 and 6fbb0595e36f9acf4f24e20e43c7c790ee91f22e